src/Extension/DefaultExtension.php

```diff
@@ -605,6 +605,98 @@ public function smarty_modifier_join($values, $separator = '')
 		return implode((string) ($separator ?? ''), (array) $values);
 	}
 
+	/**
+	 * Smarty json_encode modifier plugin.
+	 * Type:     modifier
+	 * Name:     json_encode
+	 * Purpose:  Returns the JSON representation of the given value or false on error. The resulting string will be UTF-8 encoded.
+	 *
+	 * @param mixed $value
+	 * @param int   $flags
+	 * @param string $input_encoding of $value; defaults to \Smarty\Smarty::$_CHARSET
+	 *
+	 * @return string|false
+	 */
+	public function smarty_modifier_json_encode($value, $flags = 0, string $input_encoding = null)
+	{
+		if (!$input_encoding) {
+			$input_encoding = \Smarty\Smarty::$_CHARSET;
+		}
+
+		# json_encode() expects UTF-8 input, so recursively encode $value if necessary into UTF-8
+		if ($value && strcasecmp($input_encoding, 'UTF-8')) {
+			if (is_string($value)) {	# shortcut for the most common case
+				$value = mb_convert_encoding($value, 'UTF-8', $input_encoding);
+			}
+			elseif (is_array($value) || is_object($value)) {
+				static $transcoder;	# this closure will be assigned once, and then persist in memory
+				if (is_null($transcoder)) {
```

Comment:
You probably did this because of [this comment](https://github.com/smarty-php/smarty/issues/1011#issuecomment-2099373912) but I'm afraid this is a bit much. Can you refactor the transcoder into a class file under `src/Extension/DefaultExtension/DeepTranscoder.php` or something like that?

Reply:
Alright, perhaps DeepJsonTranscode.php then since it's specific to transcoding for json_encode() as indicated by a comment in the code where JsonSerializable is checked.

Reply:
I just pushed a new commit where the json_encode() modifier uses the new class src/Extension/DefaultExtension/RecursiveTranscoder.php, and added a little to the json_encode unit tests.